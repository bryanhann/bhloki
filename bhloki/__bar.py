from collections import OrderedDict
from github import Github, GithubException

from progress.bar import Bar

from bhloki.user import User
from bhloki.util import *
from bhloki.transforms import *

LIMIT=None

class Remote:
    def __init__(self):
        self.__user = None
        self.__hub = None
        self.__repos = None
        self.__items = None

    def _user(self):
        self.__user = self.__user or User()
        return self.__user

    def _hub(self):
        self.__hub = self.__hub or self.__network_hub()
        return self.__hub

    def _repos(self):
        self.__repos = self.__repos or list(self.__network_repos())
        return self.__repos

    def _items(self):
        self.__items = self.__items or list(self.__network_items())
        return self.__items

    def dict(self):
        return OrderedDict(sorted(self._items()))

    # Always call self.user().credentials() before any Bar() to insure that
    # any userinteraction takes place outside of bars.

    def __network_hub(self):
        self._user().credentials()
        with Bar('Loading hub  ', fill='@', suffix='%(percent).1f%% - %(eta)ds') as bar:
            bar.next()
            return Github( *self._user().credentials() )

    def __network_repos(self):
        self._user().credentials()
        with Bar('Loading repos', fill='@', suffix='%(percent).1f%% - %(eta)ds') as bar:
            for repo in self._hub().get_user().get_repos():
                bar.next()
                yield repo

    def __network_items(self):
        self._user().credentials()
        with Bar('Loading shas ', fill='@', suffix='%(percent).1f%% - %(eta)ds') as bar:
            repos = self._repos()
            # Set LIMIT to a small number for testing purposes
            if LIMIT:
                repos = repos[:LIMIT]
            for repo in repos:
                bar.next()
                yield item4repo(repo)




