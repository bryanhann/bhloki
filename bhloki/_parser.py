import argparse
parser = argparse.ArgumentParser()
parser.add_argument('command',nargs='+')
parser.add_argument('--user', action='store', default='bryanhann')
parser.add_argument('--hash', action='store', default='')
parser.add_argument('--skipempty', action='store_true')
parser.add_argument('--path', action='store')
parser.add_argument('--name', action='store', default='')
ARGS  = parser.parse_args()
ARGS.args = ARGS.command[1:] + ['']*5
ARGS.cmd = ARGS.command[0]
