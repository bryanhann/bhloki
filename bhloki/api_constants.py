import os
EXCLUDE_LIST=[ '__index__' ]
TEST_MAX=os.environ.get('loki_test_max')
TEST_MAX=TEST_MAX and int(TEST_MAX)
GITHUB_PREFIX="https://github.com/bryanhann"
LOKI_DICT = os.environ.get('loki_dict', '')
GITHUB_USERNAME='bryanhann'
