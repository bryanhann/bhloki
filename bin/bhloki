#!/bin/bash
python -m pyloki $*
