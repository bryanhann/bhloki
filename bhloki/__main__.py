assert __name__ == '__main__'

from pprint import pprint
import argparse
from bhloki.util import *
from bhloki.constants import LOCAL, REMOTE
def _unzip2(pairs):
    a=[pairs[0] for pair in pairs]
    b=[pairs[1] for pair in pairs]
    return a,b
lfilter = lambda *x  : list(filter(*x))
prefix_test_4_prefix = lambda PREFIX : lambda STRING : STRING.startswith(PREFIX)
fTRUE = lambda x: True
fID = lambda x: x
def _hash():
    return ARGS.hash or die( 5, 'need to supply --hash flag' )

def filtered_dict( oDict, keytest=None, valtest=None):
    """(oDict, keytest, valtest) -> dictOfLists
    Filter a dictionary, returning a new dictionary of the same type.
    It is assumed that the dictionary values are lists.

    The keytest filters keys from the dictionary.
    The valtest filters items from the values.
    """
    for val in oDict.values(): assert type(val)==type([])
    nDict = type(oDict)()
    for key in filter( keytest, oDict.keys() ):
        nDict[key] = filter( valtest, oDict[key] )
    return nDict

def filtered4dict(
        old
        , kFilter   = fTRUE
        , vFilter   = fTRUE
        , vvMap     = fID
        , vvFilter  = fTRUE
    ):
    new = type(old)()
    for key,val in old.items():
        val = lfilter(vvFilter,lmap(vvMap,val))
        if kFilter(key) and vFilter(val):
            new[key]=val
    return new

#_value_filteredfiltered_dict_remove_empty( oDict, keytest=None, valtest=None):


COMMANDS = {}
def command(f):
    assert f.__doc__
    COMMANDS[f.__name__] = f

@command
def rlist():
    "list reverse dictionary items"
    printitems(sorted(LOCAL.rdict().items()))

@command
def _shas():
    "List all shas with the hash prefix. Requires --hash"
    for sha in sorted(LOCAL.shas4hash(_hash())):
        print(sha)

@command
def _sha():
    "Return the unique sha with the hash prefix. Requres --hash"
    pprint(LOCAL.sha4hash(_hash()))

@command
def find():
    "List the names for the given hash. Requires --hash"
    for name in LOCAL.names4sha(_hash()):
        print(name)

@command
def update():
    """Update the cached dictionary from github

    This is an interactive command that requires the user to supply credentials.

    Note: The current cached directory will be overwritten.
    """
    LOCAL.dict_write(REMOTE.dict())



@command
def _list():
    """List items in the local dictionary. The output can be filtered with --name and --hash

    The output is a list of items formatted:
        repository-name
            sha-of-latest-commit
            ...
            sha-of-second-commit
            sha-of-initial-commit

    This output can be filtered with the following options:

        --name prefix : skip repository names that do not start with prefix
        --sha  prefix : skip shas that do not start with prefix
    """

    _dict = filtered4dict(
        LOCAL.dict()
        , vvMap      = fID
        , vvFilter   = prefix_test_4_prefix(ARGS.hash)
        , kFilter    = prefix_test_4_prefix(ARGS.name)
        , vFilter    = fID if ARGS.skipempty else fTRUE
    )
    printitems( _dict.items() )


def cmd4name(name):
    try:
        return COMMANDS[name]
    except KeyError:
        die(11, "unrecognized: '%s'.  Try 'loki help'" )

@command
def help(subcommand=None):
    """print help. For help on a subcomand use 'help subcommand'"""
    if subcommand:
        doc = cmd4name(subcommand).__doc__
        print( 'Help for loki %s' % subcommand)
        print()
        print(doc)
    else:
        print('loki subcommands')
        for name, cmd in COMMANDS.items():
            doc = cmd.__doc__.split('\n')[0]
            print( '\t%s%s' % (name.ljust(10),doc) )
def main():
    name = ARGS.command[0]
    args = ARGS.command[1:]
    if not name in COMMANDS:
        die (9,  'not recognized: "%s"' % name )
    try:
        COMMANDS[name](*args)
    except UniqExc:
        die(6, 'hash either overdetermined or underdermined the sha!')

parser = argparse.ArgumentParser()
parser.add_argument('command',nargs='+')
parser.add_argument('--user', action='store', default='bryanhann')
parser.add_argument('--hash', action='store', default='')
parser.add_argument('--skipempty', action='store_true')
parser.add_argument('--path', action='store')
parser.add_argument('--name', action='store', default='')
ARGS = args = parser.parse_args()
main()
