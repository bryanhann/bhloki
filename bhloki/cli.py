# This module should be imported by [bhloki.__main__]
from pprint import pprint
from bhloki.util import *
from bhloki.constants import *
from bhloki._parser import parser
TAB = '    '
COMMANDS = dict()

def main():
    ARGS  = parser.parse_args()
    ARGS.args = ARGS.command[1:] + ['']*5
    ARGS.cmd = ARGS.command[0]
    # This function should be called by [bhloki.__main__]
    fn = COMMANDS.get(ARGS.cmd) or die(9,'unknown command: %s' % ARGS.cmd)
    try:
        fn(ARGS)
    except UniqExc:
        die(6, 'hash either overdetermined or underdermined the sha!')

def command(f):
    assert f.__doc__
    COMMANDS[f.__name__] = f

@command
def rlist(ARGS=None):
    "list reverse dictionary items"
    _pprint_items(sorted(LOCAL.rdict().items()))

@command
def _shas(ARGS=None):
    "List all shas with the hash prefix. Requires --hash"
    for sha in sorted(LOCAL.shas4hash(ARGS.hash)):
        print(sha)

@command
def _sha(ARGS=None):
    "Return the unique sha with the hash prefix. Requres --hash"
    print( LOCAL.sha4hash(ARGS.hash))

@command
def find(ARGS=None):
    "List the names for the given hash. Requires --hash"
    for name in LOCAL.names4sha(ARGS.hash):
        print(name)

@command
def update(ARGS=None):
    """Update the cached dictionary from github

    This is an interactive command that requires the user to supply credentials.

    Note: The current cached directory will be overwritten.
    """
    LOCAL.dict_write(REMOTE.dict())



@command
def _list(ARGS=None):
    """List items in the local dictionary. The output can be filtered with --name and --hash

    The output is a list of items formatted:
        repository-name
            sha-of-latest-commit
            ...
            sha-of-second-commit
            sha-of-initial-commit

    This output can be filtered with the following options:

        --name prefix : skip repository names that do not start with prefix
        --sha  prefix : skip shas that do not start with prefix
    """

    _dict = filtered4dict(
        LOCAL.dict()
        , vvMap      = fID
        , vvFilter   = prefix_test_4_prefix(ARGS.hash)
        , kFilter    = prefix_test_4_prefix(ARGS.name)
        , vFilter    = fID if ARGS.skipempty else fTRUE
    )
    _pprint_items( _dict.items() )


@command
def help(ARGS=None,*subcommand):
    """print help. For help on a subcomand use 'help subcommand'"""
    subcommand = ARGS.args[0]
    if not subcommand:
        print('loki subcommands')
        for name, cmd in COMMANDS.items():
            print( TAB*2 + name.ljust(10) + TAB + shortdoc(cmd) )
    elif subcommand in COMMANDS:
        print(COMMANDS[subcommand].__doc__)
    else:
        die(5,"unknown subcommand: '%s'" % subcommand)


# UI utilities:
def shortdoc(cmd):
    return cmd.__doc__.split('\n')[0]

def die(err, msg=''):
    msg = "[err:%s] loki: %s\n" % (err, msg)
    sys.stderr.write(msg)
    sys.stderr.flush()
    exit(err)

def _pprint_items(items):
    for k,vals in items:
        print(k)
        for val in vals:
            print('\t' + val)

