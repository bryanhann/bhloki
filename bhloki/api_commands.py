import os
import sys
from collections import OrderedDict
import argparse


from bhloki.api_constants  import *
from bhloki.api_util import *
from bhloki.api_github import *



def cmd_dump(ARGS):
    pprint_kvv_dikt( pickle_read(ARGS.path) )


def cmd_update(ARGS):
    DICTPATH=ARGS.path
    err( 'DICTPATH: ' + ARGS.path )
    old = pickle_read( ARGS.path )
    new = acquire()
    if old == new:
        die(3,'UNCHANGED')
    try:
        pickle_write(DICTPATH, new)
        assert new == pickle_read( DICTPATH  )
        die(0,'CHANGED')
    except Exception:
        raise

def cmd_url4sha(ARGS):
    ARGS.args or die(98, 'a sha must be provided')
    sha_prefix = ARGS.args[0]
    len(sha_prefix) > 4 or die(2,'too short')
    old = pickle_read( ARGS.path  ) or OrderedDict()
    for name,sha_seq in old.items():
        for sha in sha_seq:
            if sha.startswith(sha_prefix):
                print(GITHUB_PREFIX + '/' + name)
                exit()
            err(sha)
    die(1,'not found')

def parse():
    global ARGS
    parser = argparse.ArgumentParser()
    parser.add_argument('args',nargs='+')
    parser.add_argument('--user', action='store', default='bryanhann')
    parser.add_argument('--path', action='store', default=os.environ.get('loki_dict', ''))
    parser.add_argument('--hash', action='store', default='')
    parser.add_argument('--skipempty', action='store_true')
    parser.add_argument('--name', action='store', default='')
    ARGS = parser.parse_args()

def main():
    parse()
    cmdname = ARGS.args.pop(0)
    try:
        cmd = eval( 'cmd_' + cmdname )
    except:
        die(99, 'not a command: %s' % repr(cmdname))
    cmd(ARGS)
if __name__=='__main__':
    main()
