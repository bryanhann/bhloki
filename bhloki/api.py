import os
import argparse
import bhloki.api_commands as commands
import bhloki.api_constants as CC
from bhloki.api_util import die
parser = argparse.ArgumentParser()
parser.add_argument('args',nargs='+')
parser.add_argument('--user', action='store', default=CC.GITHUB_USERNAME)
parser.add_argument('--path', action='store', default=CC.LOKI_DICT)
parser.add_argument('--hash', action='store', default='')
parser.add_argument('--name', action='store', default='')
ARGS = parser.parse_args()

if __name__=='__main__':
    cmdname = ARGS.args.pop(0)
    fullname = 'cmd_%s' %  cmdname
    try:
        cmd = getattr(commands, fullname)
    except:
        die(99, 'not a command: %s' % repr(cmdname))
    cmd(ARGS)
