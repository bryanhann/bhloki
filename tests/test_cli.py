import pytest
from bhloki.cli import main
from subprocess import *

UNIQUE_HASH = "3a0"
UNIQUE_SHA  = "3a08ca9070b547bb5a95fa00ef01e1042f8e9a74"
UNIQUE_NAME = ".cache.bh.zjot"

def out_err(cmd):
    p = Popen(cmd, shell=True, bufsize=100000, stderr=PIPE, stdin=PIPE, stdout=PIPE, close_fds=True)
    out = str(p.stdout.read(), 'utf-8')
    err = str(p.stderr.read(), 'utf-8')
    return out,err

def out(cmd): return out_err(cmd)[0].strip()

def dump(cmd):
    NEWLINE='\n'
    cmd_head = '---[command]--------------------------------'
    out_head = '---[stdout]--------------------------------'
    err_head = '---[stder]--------------------------------'
    out,err = out_err(cmd)
    return NEWLINE.join( [cmd_head , cmd , out_head , out , err_head , err] )

@pytest.fixture

def loki(regtest):
    prefix = 'loki .test '
    """Usage: [run('commandline string')]. It will run and out,err will be regtested."""
    return lambda cmd: regtest.write(dump(prefix + cmd))


if 1:
    def test_help(loki):        loki( 'help' )
    def test_help_rlist(loki):  loki( 'help rlist' )
    def test_help__shas(loki):  loki( 'help _shas' )
    def test_help__sha(loki):   loki( 'help _sha' )
    def test_help_find(loki):   loki( 'help find' )
    def test_help_update(loki): loki( 'help update' )
    def test_help_list(loki):   loki( 'help _list' )
    def test_help_help(loki):   loki( 'help help' )
if 1 and "testing _list":
    L = '_list'
    S = ' --skipempty'
    N = ' --name .c'
    H = ' --hash 3'
    X = ''
    def test_l____(loki):   loki( L+X+X+X )
    def test_l___s(loki):   loki( L+X+X+S )
    def test_l__n_(loki):   loki( L+X+N+X )
    def test_l__ns(loki):   loki( L+X+N+S )
    def test_l_h__(loki):   loki( L+H+X+X )
    def test_l_h_s(loki):   loki( L+H+X+S )
    def test_l_hn_(loki):   loki( L+H+N+X )
    def test_l_hns(loki):   loki( L+H+N+S )
    def test_rlist(loki):   loki( 'rlist'  )
    def test_shas_1(loki):     loki( '_shas ' )
    def test_shas_2(loki):     loki( '_shas --hash 3'    )
    def test_shas_3(loki):     loki( '_shas --hash 3a0'  )
    def test_sha_1(loki):     loki( '_sha ' )
    def test_sha_2(loki):     loki( '_sha --hash 3'    )
    def test_sha_3(loki):     loki( '_sha --hash 3a0'  )
if 1:
    def test_sha_return_value_1():  assert UNIQUE_SHA   == out( 'loki .test _sha --hash %s' % UNIQUE_HASH)
    def test_sha_return_value_2():  assert ""           == out( 'loki .test _sha --hash xxx' )
    def test_find():                assert UNIQUE_NAME  == out( 'loki .test find --hash %s' % UNIQUE_HASH)


