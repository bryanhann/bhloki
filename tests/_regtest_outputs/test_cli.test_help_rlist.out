---[command]--------------------------------
loki .test help rlist
---[stdout]--------------------------------
list reverse dictionary items

---[stder]--------------------------------
